Add Gtest support to test Camera-vhal API's

Sapna1-singh opened the pull request:
Introduce a property based on which instance id is taken for socket connection for Gtest purpose. Check for this property in camera-vhal and if this set then use the instance id that is set by property for Gtest else use the default one.

Newly introduced property is ro.boot.container.testid.

Add Unit TestCases for Camera-vhal in Gtest.

Tracked-On: OAM-105831
Signed-off-by: Singh, Sapna1 <sapna1.singh@intel.com>

Sapna1-singh committed:
Modularize code and add a helper class for camera info validation

Modularize all the socket communication code for handling all the
incoming requests dynamically and  add a helper class that
facilitates validation of all the camera info metadata.

Tracked-On: OAM-106780
Signed-off-by: Singh, Sapna1 <sapna1.singh@intel.com>

akodanka commented on src/ClientCommunicator.cpp:
```diff
@@ -186,7 +219,12 @@ void ClientCommunicator::configureCapabilities() {
+        val_client_cap[i].validOrientation = IsSensorOrientationValid(camera_info[i].sensorOrientation);
+        val_client_cap[i].validCameraFacing = IsCameraFacingValid(camera_info[i].facing);
+
+        /*switch (camera_info[i].codec_type) {
```
You can delete these lines instead of commenting out.

akodanka commented on src/ClientCommunicator.cpp:
```diff
@@ -141,6 +172,8 @@ void ClientCommunicator::configureCapabilities() {
     if (header.type != CAMERA_INFO) {
         ALOGE("%s(%d): invalid camera_packet_type: %s", __FUNCTION__, mClientId,
               camera_type_to_str(header.type));
+        if(header.type == REQUEST_CAPABILITY)
```
Since this is not a valid usecase, I wouldn't recommend using a goto statement here.
If we really want to support the REQUEST_CAPABILITY dynamically, we should modularize all of these socket communication code, to save the current state, and handle the state changes accordingly.

akodanka commented on src/VirtualCameraFactory.cpp:
```diff
@@ -110,7 +110,10 @@ bool VirtualCameraFactory::createSocketListener() {
 
     char id[PROPERTY_VALUE_MAX] = {0};
-    if (property_get("ro.boot.container.id", id, "") > 0) {
+    if (property_get("ro.boot.container.testid", id, "") > 0) {
```
Can you leave a comment above this to indicate why this is introduced.

Sapna1-singh committed:
Add Gtest support to test Camera-vhal API's

Introduce a property based on which instance id is taken
for socket connection for Gtest purpose. Check for this
property in camera-vhal and if this set then use the
instance id that is set by property for Gtest else use
the default one.

Newly introduced property is ro.boot.container.testid.

Add Unit TestCases for Camera-vhal in Gtest.

Tracked-On: OAM-105831
Signed-off-by: Singh, Sapna1 sapna1.singh@intel.com

akodanka commented on src/ClientCommunicator.cpp:
```diff
@@ -321,16 +259,14 @@ void ClientCommunicator::configureCapabilities() {
         ALOGE("%s(%d): Failed to send camera capabilities, err: %s ", __FUNCTION__, mClientId,
               strerror(errno));
-        goto out;
+        free(ack_packet);
```
I think you can avoid multiple free calls if you put the below logs in else block, and remove this line ?

akodanka commented on src/ClientCommunicator.cpp:
```diff
@@ -399,95 +335,43 @@ bool ClientCommunicator::clientThread() {
+                            break;
+                        case CAMERA_DATA:
+                            if (!mIsConfigurationDone) {
+                                ALOGE("%s(%d): Not Valid Type.....camera_packet_type: %s, line_num = %d ", __FUNCTION__, mClientId,
```
Can you mention Configuration Incomplete in the logs to indicate why the packet is invalid.
eg. 
```
ALOGE("%s(%d): Invalid camera_packet_type: %s, Configuration not completed", __FUNCTION__, mClientId,
                                      camera_type_to_str(header.type));```

akodanka commented on src/ClientCommunicator.cpp:
```diff
@@ -399,95 +335,43 @@ bool ClientCommunicator::clientThread() {
+                                handleIncomingFrames(header.size);
+                            break;
+                        default:
+                            if (header.size > mSocketBuffer.size()) {
```
I don't think this condition should be part of the default case, instead it can be checked before entering the switch case probably.
Even if its targeted only for data packets, I think it should be handled at the beginning of `case CAMERA_DATA:` .

akodanka commented on Gtest/CameraClient.h:
```diff
@@ -0,0 +1,48 @@
+
+class CameraClient {
+private:
+    bool IsRunning;
```
I think we should have a uniform naming convention for the variables(at least for the newly introduced source files).
Could be `mIsRunning, is_running_`, but should be same across the variables used (`instance_id , video_sink` etc).

Sapna1-singh committed:
Fix crash seen in deleting VirtualCameraFactory object

ConnectionsListener class was derived from Thread class
which was causing issue while calling join for this thread
from VirtualCameraFactory destructor.

The Base class from which ConnectionsListener was derived
was not the proper class.

With current implementation, instead of deriving from thread
class, a thread is introduced as a member of the class to do
the same job and it is started during creation of
ConnectionsListener object

Tracked-On: OAM-106781
Signed-off-by: Singh, Sapna1 <sapna1.singh@intel.com>

Sapna1-singh committed:
Fix the clean up issue observed while running gtest for camera-vhal

close the client thread after completing the execution of all
testcases.

An API is introduced to do cleanup and in that API, client
thread exit is requested, socket server thread exit is
requested and joins with the main thread after all the
cleanup.

Tracked-On: OAM-106783
Signed-off-by: Singh, Sapna1 <sapna1.singh@intel.com>

Sapna1-singh commented on include/ClientCommunicatorHelper.h:
```diff
@@ -0,0 +1,40 @@
+#ifndef CLIENT_COMMUNICATOR_HELPER_H
+#define CLIENT_COMMUNICATOR_HELPER_H
+
+#include <iostream>
```
Its better to keep these 2 headers to avoid redefinition of header file issue.

Sapna1-singh commented on src/ClientCommunicator.cpp:
```diff
@@ -297,22 +233,24 @@ void ClientCommunicator::configureCapabilities() {
         // Wait till complete the metadata update for a camera.
         {
             Mutex::Autolock al(sMutex);
-            // Dynamic client configuration is not supported when camera
```
No its not required. Its already fixed from AIC application code.

Sapna1-singh committed:
Merge branch 'r/ics3a/main' into ww07_Gtest

Sapna1-singh committed:
Update ClientCommunicator.cpp
